client/asset/eth/eth.go

```diff
@@ -331,13 +361,32 @@ func (eth *ExchangeWallet) MaxOrder(lotSize uint64, feeSuggestion uint64, nfo *d
 		RealisticWorstCase: realisticWorstCase,
 		RealisticBestCase:  realisticBestCase,
 		Locked:             locked,
-	}, nil
+	}
 }
 
-// PreSwap gets order estimates based on the available funds and the wallet
-// configuration.
-func (*ExchangeWallet) PreSwap(req *asset.PreSwapForm) (*asset.PreSwap, error) {
-	return nil, asset.ErrNotImplemented
+// getInitGas gets an estimate for the gas required for initiating a swap. The
+// result is cached and reused in subsequent requests.
+func (eth *ExchangeWallet) getInitGas() (uint64, error) {
+	eth.cachedInitGasMtx.RLock()
+	cachedInitGas := eth.cachedInitGas
+	eth.cachedInitGasMtx.RUnlock()
+	if cachedInitGas != nil {
+		return *cachedInitGas, nil
+	}
+
+	eth.cachedInitGasMtx.Lock()
+	defer eth.cachedInitGasMtx.Unlock()
+
+	now := time.Now().Unix()
+	var secretHash [32]byte
+	copy(secretHash[:], encode.RandomBytes(32))
+	initGas, err := eth.node.initiateGas(eth.ctx, now, secretHash, &eth.acct.Address, &mainnetContractAddr)
+	if err != nil {
+		return 0, err
+	}
```

Comment:
Was it confirmed that a wallet without a certain balance would always fail here? I may be misremembering, but if so can check balance and fail faster. Although it looks like the places where this is called already have some balance check? If so could comment on that assumption.

Reply:
I'll add a comment.